src/main/java/com/artipie/asto/cache/StorageCache.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.cache;
+
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import com.artipie.asto.rx.RxStorageWrapper;
+import hu.akarnokd.rxjava2.interop.SingleInterop;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+/**
+ * Storage cache implementation.
+ * @since 0.24
+ */
+public final class StorageCache implements Cache {
+
+    /**
+     * Back-end storage.
+     */
+    private final Storage storage;
+
+    /**
+     * New storage cache.
+     * @param storage Back-end storage for cache
+     */
+    public StorageCache(final Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public CompletionStage<? extends Content> load(final Key key,
+        final Supplier<? extends CompletionStage<? extends Content>> remote,
+        final CacheControl control) {
+        final RxStorageWrapper rxsto = new RxStorageWrapper(this.storage);
+        return rxsto.exists(key)
+            .filter(exists -> exists)
+            .flatMapSingleElement(
+                exists -> SingleInterop.fromFuture(control.validate(key))
+                    .map(valid -> exists && valid)
+            )
+            .filter(valid -> valid)
+            .flatMapSingleElement(ignore -> rxsto.value(key))
+            .switchIfEmpty(
+                SingleInterop.fromFuture(remote.get()).flatMapCompletable(
+                    content -> rxsto.save(
+                        key, new Content.From(content.size(), content)
+                    )
```

Comment:
@g4s8 here we are waiting for content to be saved into cache before giving away the content. This makes system using cache both slower and less resilient. Maybe it would be better to do the saving in async manner and return `remote.get()` right away?

Reply:
@olegmoz yes, I just wanted to implement tests for this cache and improve it in another PR